### PR TITLE
Add filter method for ordered and unordered map

### LINKED ISSUE #67400: C++ std::map should provide a filter method
C++ dictionary types such as `std::map` and `std::unordered_map` should provide a Swift-friendly `func filter` with a similar interface to the Swift stdlib function: https://developer.apple.com/documentation/swift/dictionary/filter(_:)

This can be implemented by adding a new function to the `CxxDictionary` extension in the C++ stdlib overlay (see `stdlib/public/Cxx/CxxDictionary.swift`).

### PR DESCRIPTION
**Pull request by mohanadkandil**
Implemented a filter method like the [Swift stdlib one ](https://developer.apple.com/documentation/swift/dictionary/filter(_:))
Resolves #67400

### CODE REVIEW

**Comment by mohanadkandil** [comment]
@swift-ci Please smoke test macOS platform

**Comment by mohanadkandil** [comment]
@swift-ci Please test

**Comment by mohanadkandil** [comment]
close #67400

**Comment by finagolfin** [comment]
If you modify the original comment to add "Resolves #67400," as [shown when you submitted this pull](https://github.com/apple/swift/blob/main/.github/PULL_REQUEST_TEMPLATE.md), that issue will be linked to this pull.

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 1]
```diff
@@ -25,6 +25,22 @@ public protocol CxxDictionary<Key, Value> {
 
   /// Do not implement this function manually in Swift.
   func __endUnsafe() -> RawIterator
+
+  func filter(
+    _ isIncluded: (Key, Value) throws -> Bool
+  ) rethrows -> [Key: Value] {
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:31]: This won't compile: this function is currently defined within the protocol definition, while it should be declared in an extension for the `CxxDictionary` protocol (see `extension CxxDictionary` in the same file below)

**Comment by egorzhdan** [comment]
@mohanadkandil thanks for the PR! You won't be able to run the CI just yet, this is fine, running the CI requires write access to the repo. I can run the CI for you in the future.

Let me know if you need any help with compiling this locally on your machine.

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 1 above)
Comment (mohanadkandil) [reply stdlib/public/Cxx/CxxDictionary.swift:31]: only declaration or with it's body ?

Comment (egorzhdan) [reply stdlib/public/Cxx/CxxDictionary.swift:31]: The entire function declaration with its body should be within an extension for `CxxDictionary`. Swift doesn't require you to declare the function separately in this case:
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/protocols#Protocol-Extensions

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 2]
```diff
@@ -38,4 +38,20 @@ extension CxxDictionary {
...
+    var filteredDictionary: [Key: Value] = [:]
+    let iterator = __findUnsafe(Key)
+    let endIterator = __endUnsafe()
+
+    while iterator != endIterator {
+      let pair = iterator.pointee
+      if try isIncluded(pair.first, pair.second) {
+        filteredDictionary[pair.first] = pair.second
+      }
+      iterator.successor()
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:54]: `.successor()` doesn't mutate the object, so this call will have no effect.

**Review by egorzhdan — COMMENTED** [review]
@mohanadkandil please try building and running this on your machine

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 3]
```diff
@@ -38,4 +38,22 @@ extension CxxDictionary {
       return iter.pointee.second
     }
   }
+  
+  func filter(
+    _ isIncluded: (_ key: Key, _ value: Value) throws -> Bool
+) rethrows -> [Key: Value] {
+    var filteredDictionary: [Key: Value] = [:]
+    var iterator = __findUnsafe(Key)
```
Comment (mohanadkandil) [inline stdlib/public/Cxx/CxxDictionary.swift:46]: Is there's a way to point at the beginning of the dictionary here ?

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 2 above)
Comment (mohanadkandil) [reply stdlib/public/Cxx/CxxDictionary.swift]: I took a look at the usage of `rawIterator` and I saw that it's mutating with something like this one in `CxxSequence`
`self.rawIterator = self.rawIterator.successor()`
What do you think about this one?

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 3 above)
Comment (mohanadkandil) [reply stdlib/public/Cxx/CxxDictionary.swift:46]: @egorzhdan hey, are you still active on this PR?

Comment (egorzhdan) [reply stdlib/public/Cxx/CxxDictionary.swift:46]: Sorry for the delay!
C++ dictionaries in Swift will have `__beginUnsafe()` and `__endUnsafe()` methods (mapped from `begin()` and `end()` methods defined on the C++ side), so you can call `__beginUnsafe()` here. You'll need to declare `__beginUnsafe()` as a member of `CxxDictionary` protocol for that to work.

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 2 above)
Comment (egorzhdan) [reply stdlib/public/Cxx/CxxDictionary.swift]: Yes, this looks correct to me

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 3 above)
Comment (mohanadkandil) [reply stdlib/public/Cxx/CxxDictionary.swift]: no worries. it's done
Please check it out

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 4]
```diff
@@ -72,4 +75,22 @@ extension CxxDictionary {
       }
     }
   }
+  
+  func filter(
+    _ isIncluded: (_ key: Key, _ value: Value) throws -> Bool
+) rethrows -> [Key: Value] {
+    var filteredDictionary: [Key: Value] = [:]
+    var iterator = __beginUnsafe(Key)
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:83]: This won't compile since `__beginUnsafe` takes no parameters. Please try building this on your machine to see the compiler error, feel free to ask me questions if you run into issues when building Swift.

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 5]
```diff
@@ -29,6 +29,9 @@ public protocol CxxDictionary<Key, Value> {
   func __findUnsafe(_ key: Key) -> RawIterator
 
   /// Do not implement this function manually in Swift.
+
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:32]: Let's keep the documentation comment for `__findMutatingUnsafe`, and let's have the same comment for `__beginUnsafe` as well.

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 6]
```diff
@@ -14,7 +14,7 @@
 ///
 /// C++ standard library types such as `std::map` and `std::unordered_map`
 /// conform to this protocol.
-public protocol CxxDictionary<Key, Value> {
+public protocol CxxDictionary where Key: Hashable {
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:17]: Why did you remove primary associated types `Key, Value` from the signature? Those are important for the ergonomics of this protocol, they should stay.

**Review by egorzhdan — CHANGES_REQUESTED** [review]

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 3 above)
Comment (mohanadkandil) [reply stdlib/public/Cxx/CxxDictionary.swift]: Just built the project using this command 

`utils/build-script --skip-build-benchmarks \
  --skip-ios --skip-watchos --skip-tvos --swift-darwin-supported-archs "$(uname -m)" \
  --sccache --release-debuginfo --swift-disable-dead-stripping`
  
  but I'm just getting errors related to DAK Layouts.

Comment (egorzhdan) [reply stdlib/public/Cxx/CxxDictionary.swift]: I see. I won't be able to help without seeing the errors unfortunately. Other folks who work with `build-script` a lot should know how to fix this. Could you please post the build log you're seeing on https://forums.swift.org/c/development/compiler/9 and ask how to workaround it?

**Comment by CrazyFanFan** [comment]
I have some personal insights regarding `filter` that I would like to share. Given that `CxxDictionary` does not require its `Key` to conform to `Hashable`, I believe it would be more reasonable for the `filter` function to return `Self`. Consequently, we need a method to create a new instance of `Self` using either `std::map` or `std::unordered_map`. I have attempted to write the code for this; however, I have yet to find a suitable method for implementing the `init` function.

**Comment by egorzhdan** [comment]
> I believe it would be more reasonable for the filter function to return Self

I agree. Another reason why it should return `Self` is that there is already a way to filter a `std::map` into a Swift Dictionary: `Dictionary(myCxxMap).filter { ... }`. However, there is no way to filter a `std::map` into a `std::map`, which is what you might want to do when working with a C++ library from Swift.

**Comment by mohanadkandil** [comment]
Hi @egorzhdan, 
After posting on forums and changing the code, build errors are not related to my changes, I just tried a complete different branch and it's still getting `9 errors generated`. However, the changes above like returning self are already completed. 
So what should be the next steps ?

**Comment by mohanadkandil** [comment]
Hi @egorzhdan 
Just added a test for the filter and passed it. 

Should we run the checks ?

**Comment by egorzhdan** [comment]
@mohanadkandil awesome, let's run the CI!

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 7]
```diff
@@ -25,9 +25,11 @@ public protocol CxxDictionary<Key, Value> {
   associatedtype Size: BinaryInteger
   associatedtype InsertionResult
 
+  init()
+
   /// Do not implement this function manually in Swift.
   func __findUnsafe(_ key: Key) -> RawIterator
-
+
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:32]: This looks like an accidental change, please remove it

File: stdlib/public/Cxx/CxxDictionary.swift [hunk 8]
```diff
@@ -72,4 +77,22 @@ extension CxxDictionary {
       }
     }
   }
+  
+public func filter(_ isIncluded: (_ key: Key, _ value: Value) throws -> Bool) rethrows -> Self {
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxDictionary.swift:81]: Please adjust indentation on this line

File: test/Interop/Cxx/stdlib/use-std-map.swift [hunk 9]
```diff
@@ -88,6 +88,30 @@ StdMapTestSuite.test("UnorderedMap.subscript") {
...
+StdMapTestSuite.test("Map.filter") {
+  var m = initMap()
+
+  expectNotNil(m[1])
+
+  m = m.filter { k, v in k != 1 }
+
+  expectNil(m[1])
+  expectEqual(m[2], 2)
+  expectEqual(m[3], 3)
```
Comment (egorzhdan) [inline test/Interop/Cxx/stdlib/use-std-map.swift:100]: Could you please also add a test to ensure that calling `.filter` on an empty dictionary produces another empty dictionary?

**Review by egorzhdan — COMMENTED** [review]
This mostly looks good to me, thanks @mohanadkandil!
Left a couple comments.

File: test/Interop/Cxx/stdlib/use-std-map.swift (hunk 9 above)
Comment (mohanadkandil) [reply test/Interop/Cxx/stdlib/use-std-map.swift:103]: Added tests for `map` and `unordered_map`.
If you have any comments on the empty dictionary test approach, please let me know.

File: test/Interop/Cxx/stdlib/use-std-map.swift [hunk 10]
```diff
@@ -88,6 +88,38 @@ StdMapTestSuite.test("UnorderedMap.subscript") {
   expectNil(m[-1])
 }
 
+StdMapTestSuite.test("Map.filter") {
+  var m = initMap()
+  var n: [Int: Int] = [:]
```
Comment (egorzhdan) [inline test/Interop/Cxx/stdlib/use-std-map.swift:93]: `[Int: Int]` is a Swift Dictionary type. It already has a `filter` method and is tested elsewhere. This test (and this patch) is specific to C++ dictionaries (`std::map` and `std::unordered_map`), so we should be testing C++ dictionaries here.

Comment (mohanadkandil) [reply test/Interop/Cxx/stdlib/use-std-map.swift:93]: Yeah, noticed that. 
I tested it out with `initMap() `and `initUnorderedMap()`, but I'm getting 
```
stdout>>> first:  3 (of type Swift.Int)
stdout>>> second: 0 (of type Swift.Int)
``` in test results

Comment (Obverser) [reply test/Interop/Cxx/stdlib/use-std-map.swift:93]: To get an empty dictionary in the C++ form, you're going to either need to change the header file to have initMapEmpty() or initUnorderedMapEmpty(). Alternatively, you could run a filter on a initMap() such that no keys match (aka, false literal), and then test on that new map.

Comment (Obverser) [reply test/Interop/Cxx/stdlib/use-std-map.swift:93]: Here's where the header file is:
[header for map inputs](test/Interop/Cxx/stdlib/Inputs/std-map.h)

Comment (mohanadkandil) [reply test/Interop/Cxx/stdlib/use-std-map.swift]: Thanks for your recommendation @Obverser, 
InitMap() was retuning a dictionary with 3 pairs (I saw it also from the [header](https://github.com/apple/swift/blob/main/test/Interop/Cxx/stdlib/Inputs/std-map.h)).

It's now working as expected 👍

File: stdlib/public/Cxx/CxxDictionary.swift (hunk 8 above)
Comment (egorzhdan) [reply stdlib/public/Cxx/CxxDictionary.swift:79]: Looks like the formatting isn't quite correct, let's have a two-space indentation and a blank line before this method

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by egorzhdan — APPROVED** [review]
This looks great to me now, thank you @mohanadkandil!

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS

**Comment by egorzhdan** [comment]
@swift-ci please smoke test Linux

**Comment by egorzhdan** [comment]
I'm going to merge this now. @mohanadkandil next time could you please squash all your commits in a PR into one commit 😉

**Comment by mrousavy** [comment]
Sorry this is maybe the wrong place to ask this, but does Swift's umbrella header expose a `swift::Dictionary<K, V>` type, or how should passing around maps/dictionaries between Swift <-> C++ be handled?

**Comment by egorzhdan** [comment]
> does Swift's umbrella header expose a swift::Dictionary<K, V> type

Not at the moment, no.

> how should passing around maps/dictionaries between Swift <-> C++ be handled

You could pass around an instance of `std::map`, that way you could access it from both Swift and C++.
